### ISSUE TITLE
Choosing a tiling algorithm

### ISSUE THREAD

**Issue by levy04** [Gilgames32/printf#4]
## Dynamic programming 
In [this paper](https://www.sciencedirect.com/science/article/pii/S0307904X05000314#aep-section-id25) Yaodong Cui shows 4 algorithms for different constraints. These algorithms produce the simplest optimal cutting pattern. Their time complexity unnknown.

### Tarnowski et al.
The above paper also mentions two other papers, one is [the original by Tarnowski et al.](https://www.tandfonline.com/doi/abs/10.1080/03155986.1994.11732257) which produces an optimal solution in polyinomial time, but it isn't the simplest to cut.

From the paper by Yaodong Cui:
> A polynomial time algorithm was introduced in [[10]](https://www.tandfonline.com/doi/abs/10.1080/03155986.1994.11732257) for the unconstrained problem. The time efficiency of the algorithm is high. The patterns generated are slightly difficult to cut since they are often not the simplest ones.

### Arslanov
This paper by Arslanov could also be of interest. [It details a _linear_ algorithm for the unconstrained case](https://www.sciencedirect.com/science/article/abs/pii/S0377221799004543).

From the paper by Yaodong Cui:
> Arslanov constructed a linear time algorithm for the unconstrained problem in [[11]](https://www.sciencedirect.com/science/article/abs/pii/S0377221799004543). The algorithm has the highest time efficiency. Similarly it does not guarantee that the patterns generated are the simplest ones for cutting.

## Questions
Which is more important for our use case, time-complexity, or ease of cutting after printing?

**Comment by levy04** [comment]
# Specification of problem

## Single rectangle

After more research, the [guillotine cutting](https://en.wikipedia.org/wiki/Guillotine_cutting) wiki page contains this:

> In the guillotine strip cutting problem, the large rectangle has infinite height (but a fixed width), and the goal is to cut a single rectangle of each type, such that the total material used (equivalently, the total height) is minimized. It is a variant of the two-dimensional [Strip packing problem](https://en.wikipedia.org/wiki/Strip_packing_problem).

### Strip packing
TL;DR (from the [wiki page](https://en.wikipedia.org/wiki/Strip_packing_problem))
> The strip packing problem is a 2-dimensional geometric minimization problem. Given a set of axis-aligned rectangles and a strip of bounded width and infinite height, determine an overlapping-free packing of the rectangles into the strip, minimizing its height. 

This seems to be the exact problem we face. It's a strongly-NP hard problem, but there are various polynomial time algorithms that produce an approximation. The [wiki page](https://en.wikipedia.org/wiki/Strip_packing_problem) has a couple interesting algorithms that don't seem too complicated to implement.

#### [Reverse-fit (RF)](https://en.wikipedia.org/wiki/Strip_packing_problem#Reverse-fit_(RF))
This algorithm is quite simple, and has an approximation ratio of 2, which I don't think we need to go lower than.

> Finding an algorithm with a ratio below $2$ seems complicated, and the complexity of the corresponding algorithms increases regarding their running time and their descriptions.

#### [Steinberg's algorithm (ST)](https://en.wikipedia.org/wiki/Strip_packing_problem#Steinberg's_algorithm_(ST))

This algorithm has the same approximation ratio as Reverse-Fit, but it's a _recursive_ algorithm, so we should at least look into the performance of it, and compare it to Reverse-Fit.

#### :notes: Can we get much ~~higher~~ lower? :notes: 

In the case that these algorithms produce too much waste and performance ~~(and sanity)~~ isn't an issue, we could look into algorithm's with lower approximation ratios. The current lowest is $(\frac{5}{3} + \epsilon)$, but implementing it is probably too much for a "hobby project"...
